### PR TITLE
[Lockfile Explorer] Display Connection Status Between Client and Server

### PR DESCRIPTION
**Pull request by william2958**
## Summary
This PR adds the ability for the client & server to detect if either has been disconnected, which will provide a better user experience so users don't try to accidentally open up multiple instances of either the client or server.

## Details
The error message shown if the server has not heard from the client in a while:
image

The server will also show a reconnection message after the client is re-opened
image

The error message shown if the client cannot contact the server:
image

The button to re-connect is mostly aesthetic and for better UX (Although it does try to manually reconnect), as the application automatically attempts to reconnect every 2 seconds in the background. It would probably be a bad experience to show a unpassable modal with no user interaction components.

### CODE REVIEW

File: apps/lockfile-explorer-web/src/components/ConnectionModal/index.tsx [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+  const checkAliveManual = useCallback(() => {
+    setChecking(true);
+    setManualChecked(true);
+    keepAlive().catch((e) => {
+      // Keep alive cannot fail
+    });
+  }, []);
+
+  if (isAlive) return null;
```
Comment (iclanton) [inline apps/lockfile-explorer-web/src/components/ConnectionModal/index.tsx:36]:
```suggestion
  if (isAlive) {
    return null;
  }
```

File: apps/lockfile-explorer-web/src/components/ConnectionModal/index.tsx [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  }
+
+  useEffect(() => {
+    window.setInterval(keepAlive, 2000);
+  }, []);
+
+  const checkAliveManual = useCallback(() => {
+    setChecking(true);
+    setManualChecked(true);
+    keepAlive().catch((e) => {
```
Comment (iclanton) [inline apps/lockfile-explorer-web/src/components/ConnectionModal/index.tsx:31]: Just suppress the ESLint rule.

File: apps/lockfile-explorer/package.json [hunk 3]
```diff
@@ -50,6 +50,7 @@
     "express": "4.18.1",
     "js-yaml": "~3.13.1",
     "open": "~8.4.0",
-    "cors": "~2.8.5"
+    "cors": "~2.8.5",
+    "chalk": "~4.1.2"
```
Comment (iclanton) [inline apps/lockfile-explorer/package.json:54]: Rush uses `colors`.

File: common/config/rush/browser-approved-packages.json [hunk 4]
```diff
@@ -18,6 +18,10 @@
       "name": "axios",
       "allowedCategories": [ "libraries" ]
     },
+    {
+      "name": "chalk",
+      "allowedCategories": [ "libraries" ]
+    },
```
Comment (iclanton) [inline common/config/rush/browser-approved-packages.json:24]: Use `colors` (imported as `colors/safe`), but this should go in `nonbrowser-approved-packages.json`

Comment (iclanton) [reply common/config/rush/browser-approved-packages.json:24]: Or, you can use `Terminal` from `node-core-library`, and use its `Colors` functions.

Comment (octogonz) [reply common/config/rush/browser-approved-packages.json:24]: +1

File: apps/lockfile-explorer-web/src/components/ConnectionModal/index.tsx (hunk 2 above)
Comment (octogonz) [reply apps/lockfile-explorer-web/src/components/ConnectionModal/index.tsx:31]: @iclanton I disagree with this practice.  Perhaps `keepAlive()` cannot fail today, but in the future some person may come along and modify that function, unaware of the assumption here that it's not supposed to throw.  In its current form already, to prove that `keepAlive()` cannot fail we must inspect an external source file to analyze the `checkAlive()` function as well.

The safest practice is to add the catch block that logs if the assumption is violated (e.g. `console.error('Unexpected exception: ' + e.toString());`).  The savings of omitting such safeguards really aren't justified.

I would go further to say that `no-floating-promises` lint rule is one that should /never/ be suppressed.  A couple weeks ago I met a senior engineer for coffee to help him debug a mysterious promise failure, whose ultimate cause was that someone else had mistakenly disabled this lint rule in one file. 😆

File: apps/lockfile-explorer-web/src/components/ConnectionModal/index.tsx (hunk 1 above)
Comment (william2958) [reply apps/lockfile-explorer-web/src/components/ConnectionModal/index.tsx]: changed

File: apps/lockfile-explorer-web/src/components/ConnectionModal/index.tsx (hunk 2 above)
Comment (william2958) [reply apps/lockfile-explorer-web/src/components/ConnectionModal/index.tsx:31]: Added the console.error message

File: apps/lockfile-explorer-web/src/components/ConnectionModal/index.tsx [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    <div className={styles.DisconnectOverlayBackground}>
+      <div className={`${styles.DisconnectOverlay} ${appStyles.ContainerCard}`}>
+        <h5>The Server Has Disconnected!</h5>
+        {manualChecked ? (
+          <p>We were still not able to establish a connection to the server, are you sure it is running?</p>
+        ) : (
+          <p>Please re-start the local development server to continue using this application.</p>
+        )}
+        <div className={styles.DisconnectCheckRow}>
+          <button onClick={checkAliveManual}>Check Again</button>
```
[the remark this answers is not in the document]
Comment (william2958) [reply apps/lockfile-explorer-web/src/components/ConnectionModal/index.tsx]: fixed

File: apps/lockfile-explorer/package.json (hunk 3 above)
Comment (william2958) [reply apps/lockfile-explorer/package.json]: changed

File: common/config/rush/browser-approved-packages.json (hunk 4 above)
Comment (william2958) [reply common/config/rush/browser-approved-packages.json]: changed to use colors/safe

**Review by iclanton — APPROVED** [review]

**Review by octogonz — APPROVED** [review]

**Pull request merged**